src/test/controller/UserController.test.ts

```diff
@@ -1,30 +1,52 @@
-import { getManager, getCustomRepository } from 'typeorm';
+import { getManager, getCustomRepository, EntityManager } from 'typeorm';
 import { User } from '../../main/entity/User';
 import { UserRepository } from '../../main/repository/UserRepository';
 import connection from '../repository/BaseRepositoryTest';
 import { Express } from 'express';
 import request from 'supertest';
 import { UserFactory } from '../factory/UserFactory';
-import { connectRedisClient, registerTestApplication } from './BaseController';
+import {
+  connectRedisClient,
+  registerTestApplication,
+  createFirebaseMock,
+} from './BaseController';
 import { PromotionFactory } from '../factory/PromotionFactory';
 import { DiscountFactory } from '../factory/DiscountFactory';
 import { ScheduleFactory } from '../factory/ScheduleFactory';
 import { PromotionRepository } from '../../main/repository/PromotionRepository';
-import { SavedPromotionRepository } from '../../main/repository/SavedPromotionRepository';
-import { SavedPromotionFactory } from '../factory/SavedPromotionFactory';
 import { RedisClient } from 'redis-mock';
+import * as dotenv from 'dotenv';
+import { SavedPromotion } from '../../main/entity/SavedPromotion';
+import { Promotion } from '../../main/entity/Promotion';
+dotenv.config();
 
 describe('Unit tests for UserController', function () {
   let userRepository: UserRepository;
   let promotionRepository: PromotionRepository;
-  let savedPromotionRepository: SavedPromotionRepository;
   let app: Express;
   let redisClient: RedisClient;
+  let mockFirebaseAdmin: any;
+  const uid = 'test-uid';
+  let idToken = '';
 
   beforeAll(async () => {
     await connection.create();
     redisClient = await connectRedisClient();
-    app = await registerTestApplication(redisClient);
+
+    // init mock firebase
+    mockFirebaseAdmin = createFirebaseMock();
+
+    app = await registerTestApplication(redisClient, mockFirebaseAdmin);
+
+    mockFirebaseAdmin.autoFlush();
+
+    // create user
+    const user = await mockFirebaseAdmin.createUser({
+      uid: uid,
+      email: 'test@gmail.com',
+      password: 'testpassword',
+    });
```

Comment:
Ah so what's happening here is essentially we are doing the `User sign up` logic in the FE, similar to your doc?

Reply:
yes, actually, FE sign up/in using email and password.